### PR TITLE
bpf: Reply NA when recv ND for local IPv6 endpoints

### LINKED ISSUE #10935: Passthrough NDP for IPv6
## Proposal / RFE

**Is your feature request related to a problem?** Yes

**Describe the problem you'd like**

```
# Pod's IPv6 address is 2620:0:1000:2630:2::ebb9, and it's not responding the following NS 
# request, because cilium node didn't pass the request thru. As a result,  the traffic to the pod 
# cannot be delivered.
2620:0:1000:2630:250:56ff:fe82:eec6 > ff02::1:ff00:ebb9: ICMP6, neighbor solicitation, who has 2620:0:1000:2630:2::ebb9, length 32

# I tried the following config, but it seems not working.
sysctl net.ipv6.conf.all.proxy_ndp=1
```

## Cilium configs
cilium version: v1.7.2
k8s version: v1.18.1 
```
cat /etc/cni/net.d/05-cilium.conf
{
  "cniVersion": "0.3.1",
  "name": "cilium",
  "type": "cilium-cni",
  "enable-debug": false
}
```

### PR DESCRIPTION
**Pull request by anfernee**
When receiving ICMPv6 ND message from native NIC, lookup local endpoints
for target IPv6 address. If local endpoint was found, reply NA message
via the same NIC.

This enables pod's IPv6 address discoverable on the subnet to which the
node is connected to.

Torwards #10935
Signed-off-by: Yongkun Gui <ygui@google.com>

Please ensure your pull request adheres to the following guidelines:

- [ x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #10935

```release-note
Add ND responder on native NIC to make endpoint IPv6 address discoverable on node network.
```

TODO: test still to be added.

### CODE REVIEW

File: bpf/init.sh [hunk 1]
```diff
@@ -672,6 +681,18 @@ for iface in $(ip -o -a l | awk '{print $2}' | cut -d: -f1 | cut -d@ -f1 | grep
...
 
+case "${MODE}" in
+	"flannel")
+		;;
+	"ipvlan")
+		;;
+	*)
+		COPTS=""
+		CALLS_MAP="cilium_calls_netdev_ns_${ID_HOST}"
+		bpf_load $NATIVE_DEVS "$COPTS" "ingress" bpf_netdev.c bpf_netdev.o from-netdev $CALLS_MAP
```
Comment (pchaigno) [inline bpf/init.sh]: I don't think adding a new BPF program is the right way to fix this. We already attach BPF programs to native devices from Golang (cf. [`pkg/datapath/loader/loader.go`](https://github.com/cilium/cilium/blob/c216dfb3ad911ca5ba7659b6266109c8c6882d70/pkg/datapath/loader/loader.go#L195)), so this will overwrite them. Doing so will also bypass the host firewall and break NodePort.

Instead, we should probably extend `icmp6_handle()`, which is already called from `bpf_host`:
https://github.com/cilium/cilium/blob/c216dfb3ad911ca5ba7659b6266109c8c6882d70/bpf/bpf_host.c#L370

Please see #12049 though. The `icmp6_handle()` call in `bpf_host` is currently broken (not compiled in) and the first commit in that pull request fixes it. The bug was introduced in v1.8 and we will backport that PR there.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Why is passing ICMPv6 ND packets to the stack not enough? Is it because these packets may be dropped by pod ingress policies?

File: bpf/init.sh (hunk 1 above)
Comment (anfernee) [reply bpf/init.sh]: I noticed `bpf_host` is broken, too. glad that you are fixing it. I'll look into #12049 too. 

I also updated the `icmp6_handle()`. My intention is for the node to respond to the NS msg from external so I attached it on `native_dev`.  If I have questions, I'll ping you on slack.

Comment (pchaigno) [reply bpf/init.sh]: If you try #12049 out, it might address your concern. In its current state, it will fix `bpf_host` to let all ICMPv6 NS/NA messages up the stack (in accordance with RFC4890).

File: contrib/vagrant/start.sh [hunk 2]
```diff
@@ -335,6 +335,7 @@ function write_cilium_cfg() {
     fi
 
     cilium_options+=" ${TUNNEL_MODE_STRING}"
+    cilium_options+=" --device enp0s8"
```
Comment (aanm) [inline contrib/vagrant/start.sh]: Why this change? the `device` option has been deprecated.

**Review by aanm — DISMISSED** [review]

File: bpf/init.sh (hunk 1 above)
Comment (anfernee) [reply bpf/init.sh]: I've tried #12049, the pods won't receive NS message, because the [destination IP](https://en.wikipedia.org/wiki/Solicited-node_multicast_address) is not the pod IP. I still need this but I will rebase it on top of #12049 . I will remove the new ebpf program and reuse `bpf_host`.

File: contrib/vagrant/start.sh (hunk 2 above)
Comment (anfernee) [reply contrib/vagrant/start.sh]: Ah, It should be `--devices`..

**Comment by aanm** [comment]
test-me-please

**Comment by anfernee** [comment]
> Why is passing ICMPv6 ND packets to the stack not enough? Is it because these packets may be dropped by pod ingress policies?

I think I replied this, but I cannot find it. probably github folded it, I don't know.
The reason is that the ND packet's destination address is a multicast address, which cilium doesn't recognize. The ebpf program could look into the packet and check the target address.

**Comment by anfernee** [comment]
A friendly ping :)

File: bpf/lib/icmp6.h [hunk 3]
```diff
@@ -154,8 +155,17 @@ static __always_inline int icmp6_send_echo_reply(struct __ctx_buff *ctx,
 	return DROP_MISSED_TAIL_CALL;
 }
 
+/*
+ * send_icmp6_ndisc_adv
+ * @ctx:	   socket buffer
+ * @nh_off:	   offset to the IPv6 header
+ * @mac:       device mac address
+ * @to_router: ndisc is sent to router, otherwise ndisc is sent to an endpoint.
```
Comment (qmonnet) [inline bpf/lib/icmp6.h]: Nit: alignment differ between the first two and the last two arguments.

File: bpf/lib/icmp6.h [hunk 4]
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
...
 	BPF_V6(router, ROUTER_IP);
 	if (ipv6_addrcmp(&target, &router) == 0) {
 		union macaddr router_mac = NODE_MAC;
 
-		return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac);
+		return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac, true);
+	} else if (ep) {
+		union macaddr node_mac = NODE_MAC;
+
+		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, false);
```
Comment (qmonnet) [inline bpf/lib/icmp6.h]:
```suggestion
		union macaddr node_MAC = NODE_MAC;
		
		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, !ep);
```

File: bpf/lib/overloadable_skb.h [hunk 5]
```diff
@@ -73,9 +73,16 @@ redirect_self(const struct __sk_buff *ctx)
 	 */
 #ifdef ENABLE_HOST_REDIRECT
 	return redirect(ctx->ifindex, 0);
+#else
+#ifdef NATIVE_IFINDEX
```
Comment (qmonnet) [inline bpf/lib/overloadable_skb.h]: `#elif NATIVE_IFINDEX`?

File: bpf/lib/eps.h [hunk 6]
```diff
@@ -10,16 +10,22 @@
 #include "maps.h"
 
 static __always_inline __maybe_unused struct endpoint_info *
-lookup_ip6_endpoint(struct ipv6hdr *ip6)
+__lookup_ip6_endpoint(const union v6addr *ip6)
```
Comment (qmonnet) [inline bpf/lib/eps.h]: Could we find another function name? The leading `__` tends to make me believe this function shouldn't be called from outside of `lookup_ip6_endpoint()`.

Maybe something along `lookup_ip6_endpoint_from_daddr()`?

**Review by qmonnet — CHANGES_REQUESTED** [review]
Couple of nits/suggestions from my side. @pchaigno, was your concern addressed?

**Comment by anfernee** [comment]
@qmonnet thanks for the review. new patch applied.

File: bpf/lib/icmp6.h (hunk 4 above)
Comment (qmonnet) [reply bpf/lib/icmp6.h]: Woops, I'm sorry, I misread the conditions on this block and made a bad suggestion here :(.

What I had in mind was to condense the whole `if`/`else if` block. Reading again, it could be written like this:
```c
bool router_target = ipv6_addrcmp(&target, &router);

if (router_target || ep) {
	union macaddr node_mac = NODE_MAC;

	return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, router_target);
}
```

Otherwise, your initial version was fine, too. But my suggestion of passing `!ep` to `send_icmp6_ndisc_adv()` on the two branches was wrong (you did it just for the second one which at least is technically correct, but not what we should do). Apologies about that. Would you mind changing again, please? (Either this suggestion or your original version, as you prefer.)

File: bpf/lib/icmp6.h [hunk 7]
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
...
 	union v6addr target, router;
+	struct endpoint_info *ep;
 
 	if (ctx_load_bytes(ctx, nh_off + ICMP6_ND_TARGET_OFFSET, target.addr,
 			   sizeof(((struct ipv6hdr *)NULL)->saddr)) < 0)
 		return DROP_INVALID;
 
 	cilium_dbg(ctx, DBG_ICMP6_NS, target.p3, target.p4);
 
+	ep = lookup_ip6_endpoint_from_daddr(&target);
```
Comment (pchaigno) [inline bpf/lib/icmp6.h]: Should we only perform this lookup in the else case (i.e., `ipv6_addrcmp(&target, &router) != 0`)?

File: bpf/lib/icmp6.h [hunk 8]
```diff
@@ -171,9 +181,16 @@ static __always_inline int send_icmp6_ndisc_adv(struct __ctx_buff *ctx,
...
-	icmp6hdr.icmp6_solicited = 1;
-	icmp6hdr.icmp6_override = 0;
+
+	if (to_router) {
+		icmp6hdr.icmp6_router = 1;
+		icmp6hdr.icmp6_solicited = 1;
+		icmp6hdr.icmp6_override = 0;
+	} else {
+		icmp6hdr.icmp6_router = 0;
+		icmp6hdr.icmp6_solicited = 0;
```
Comment (pchaigno) [inline bpf/lib/icmp6.h]: Shouldn't this flag be 1 since we're answering an NS message?

File: bpf/lib/icmp6.h [hunk 9]
```diff
@@ -171,9 +181,16 @@ static __always_inline int send_icmp6_ndisc_adv(struct __ctx_buff *ctx,
...
 	icmp6hdr.icmp6_cksum = icmp6hdr_old.icmp6_cksum;
 	icmp6hdr.icmp6_dataun.un_data32[0] = 0;
-	icmp6hdr.icmp6_router = 1;
-	icmp6hdr.icmp6_solicited = 1;
-	icmp6hdr.icmp6_override = 0;
+
+	if (to_router) {
+		icmp6hdr.icmp6_router = 1;
+		icmp6hdr.icmp6_solicited = 1;
+		icmp6hdr.icmp6_override = 0;
```
Comment (pchaigno) [inline bpf/lib/icmp6.h:188]: Not related to this PR, but I think this bit should be set. From the [RFC](https://tools.ietf.org/html/rfc4861#page-23):
```
                                It SHOULD NOT be
set in solicited advertisements for anycast
addresses and in solicited proxy advertisements.
It SHOULD be set in other solicited advertisements
and in unsolicited advertisements.
```

File: bpf/lib/icmp6.h [hunk 10]
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
...
 	BPF_V6(router, ROUTER_IP);
 	if (ipv6_addrcmp(&target, &router) == 0) {
 		union macaddr router_mac = NODE_MAC;
 
-		return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac);
+		return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac, true);
+	} else if (ep) {
+		union macaddr node_mac = NODE_MAC;
+
+		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, !ep);
```
Comment (pchaigno) [inline bpf/lib/icmp6.h]:
```suggestion
		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, false);
```

File: bpf/lib/overloadable_skb.h [hunk 11]
```diff
@@ -73,6 +73,11 @@ redirect_self(const struct __sk_buff *ctx)
 	 */
 #ifdef ENABLE_HOST_REDIRECT
 	return redirect(ctx->ifindex, 0);
+#elif NATIVE_IFINDEX
+	if (ctx->ifindex == NATIVE_IFINDEX)
+		return redirect(ctx->ifindex, 0);
+	else
+		return redirect(ctx->ifindex, BPF_F_INGRESS);
```
Comment (pchaigno) [inline bpf/lib/overloadable_skb.h]: Why is this required?

File: bpf/lib/icmp6.h [hunk 12]
```diff
@@ -428,6 +452,10 @@ icmp6_host_handle(struct __ctx_buff *ctx __maybe_unused)
 {
 	__u8 type __maybe_unused;
 
+	type = icmp6_load_type(ctx, ETH_HLEN);
+	if (type == ICMP6_NS_MSG_TYPE)
+		return icmp6_handle_ns(ctx, ETH_HLEN, METRIC_INGRESS);
```
Comment (pchaigno) [inline bpf/lib/icmp6.h:459]: In this case, we will bypass the host firewall, but that's okay since that's what we do for NS messages anyway. However, could you switch the comment `CTX_ACT_OK` in the table below to `icmp6_handle_ns` for `ICMPv6-neighbor-solicit`?

**Review by pchaigno — DISMISSED** [review]
> The reason is that the ND packet's destination address is a multicast address, which cilium doesn't recognize. The ebpf program could look into the packet and check the target address.

I'm not very familiar with these ICMPv6 mechanisms, so I'm probably just missing something, but I don't understand why the Linux stack can't handle that for us? Couldn't we teach it about pod IPv6 addresses so that it's able to answer NS messages to the node?

File: bpf/lib/overloadable_skb.h (hunk 11 above)
Comment (anfernee) [reply bpf/lib/overloadable_skb.h]: trying to keep the old behavior as much as possible. I am worried to break old logic.

File: bpf/lib/icmp6.h (hunk 7 above)
Comment (anfernee) [reply bpf/lib/icmp6.h]: The code is changed to single branch, how about this:
```
        ep = lookup_ip6_endpoint_from_daddr(&target);

        BPF_V6(router, ROUTER_IP);
        router_target = ipv6_addrcmp(&target, &router);

        if (router_target || ep) {
                union macaddr router_mac = NODE_MAC;

                return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac, router_target);
        }

```

Comment (pchaigno) [reply bpf/lib/icmp6.h]: My main concern is that we're doing a map lookup that's unnecessary when `router_target` is true.

File: bpf/lib/overloadable_skb.h (hunk 11 above)
Comment (pchaigno) [reply bpf/lib/overloadable_skb.h]: I mean why is any change to this file required? You are not even calling `redirect_self()`, right?

Comment (anfernee) [reply bpf/lib/overloadable_skb.h]: if does call `redirect_self()`: `send_icmp6_ndisc_adv => icmp6_send_reply => redirect_self`

**Review by qmonnet — APPROVED** [review]
Good from my side, thanks.

**Comment by anfernee** [comment]
test-me-please

File: bpf/lib/overloadable_skb.h (hunk 11 above)
Comment (pchaigno) [reply bpf/lib/overloadable_skb.h]: I still don't understand why this change is needed.
1. I couldn't find any definition of `NATIVE_IFINDEX` in the source code. Where is it defined? Should it be `NATIVE_DEV_IFINDEX`?
2. Assuming `NATIVE_IFINDEX` exists and is the index of a native device, why do we need to redirect packets to the _ingress_ of that device?

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: bpf/lib/overloadable_skb.h (hunk 11 above)
Comment (anfernee) [reply bpf/lib/overloadable_skb.h]: Originally it was only doing `redirect(ctx->ifindex, BPF_F_INGRESS);`, actually I don't understand which code path will hit this. I am guessing it could be `cilium_host` and redirect the packet back to the pods. Now the patch basically added a direction. If the packet is coming from native interface, redirect as egress, otherwise still ingress.

Comment (anfernee) [reply bpf/lib/overloadable_skb.h]: This is not useful after rebasing on master. I've removed it.

**Review by pchaigno — APPROVED** [review]

File: bpf/lib/icmp6.h [hunk 13]
```diff
@@ -355,10 +374,19 @@ static __always_inline int __icmp6_handle_ns(struct __ctx_buff *ctx, int nh_off)
 	cilium_dbg(ctx, DBG_ICMP6_NS, target.p3, target.p4);
 
 	BPF_V6(router, ROUTER_IP);
-	if (ipv6_addrcmp(&target, &router) == 0) {
+	router_target = ipv6_addrcmp(&target, &router);
+
+	if (router_target) {
```
Comment (joestringer) [inline bpf/lib/icmp6.h]: Isn't this condition the opposite of what the previous code tested for?

**Review by joestringer — CHANGES_REQUESTED** [review]
One conditional below seems to be swapped to the opposite value.

File: bpf/lib/icmp6.h (hunk 13 above)
Comment (anfernee) [reply bpf/lib/icmp6.h]: Ah.. sorry, it is somehow mistakenly changed.

**Comment by pchaigno** [comment]
test-me-please

**Comment by qmonnet** [comment]
retest-net-next

**Comment by anfernee** [comment]
@joestringer is this change okay  :)

File: bpf/lib/icmp6.h [hunk 14]
```diff
@@ -355,10 +374,19 @@ static __always_inline int __icmp6_handle_ns(struct __ctx_buff *ctx, int nh_off)
 	cilium_dbg(ctx, DBG_ICMP6_NS, target.p3, target.p4);
 
 	BPF_V6(router, ROUTER_IP);
-	if (ipv6_addrcmp(&target, &router) == 0) {
+	router_target = ipv6_addrcmp(&target, &router);
+
+	if (!router_target) {
```
Comment (joestringer) [inline bpf/lib/icmp6.h]: nit: Is there any reason to change these lines from the previous version?

`ipv6_addrcmp()` returns a u64, which is a bit unusual to convert into a bool. OTOH, the existing style of performing a comparison and checking against `0` is fairly standard compare logic in C.

**Review by joestringer — CHANGES_REQUESTED** [review]
One more minor nit, not sure the full impact of the conversion of `u64` to `bool` but IMO the previous version of the below statement was easier to read with the prior code.

With that changed we're good to merge.

File: bpf/lib/icmp6.h (hunk 14 above)
Comment (joestringer) [reply bpf/lib/icmp6.h]: I might add here, this condition is actually the case where the target is the same as the router IP so checking `!router_target` implies that this is the case that *doesn't* handle the router target, but given the final bool passed to `send_icmp6_ndisc_adv()` below, we can infer that this really is the path that handles `to_router`. So probably worth clearing up this potential confusion for future readers.

Comment (anfernee) [reply bpf/lib/icmp6.h]: I think I copied from 
https://github.com/cilium/cilium/blob/8a03d54da11a1c15302f4f90de8bf1e539252784/bpf/lib/icmp6.h#L417

but I am you are right adding a `bool` here is not necessary. I will change it back to original shape.

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
this test failed, but I couldn't see why it failed.. `21:10:49  [Fail] K8sChaosTest Connectivity demo application [AfterEach] Endpoint can still connect while Cilium is not running `

**Comment by joestringer** [comment]
test-focus K8sChaosTest.*demo

**Comment by anfernee** [comment]
test-focus K8sChaosTest.*

**Comment by anfernee** [comment]
test-focus K8sChaos

**Comment by anfernee** [comment]
test-gke

**Comment by joestringer** [comment]
@anfernee What's happening with these builds? Are they failing in the same way on multiple occasions?

**Comment by anfernee** [comment]
No, it was not triggered. `test-gke` triggered the retest..

**Comment by pchaigno** [comment]
@anfernee All builds are green, but I noticed `K8sChaosTest` tests were re-triggered a couple times without new push since then. Do we have a flake there? Is it a known flake?

**Comment by anfernee** [comment]
I saw this error: `21:10:49 [Fail] K8sChaosTest Connectivity demo application [AfterEach] Endpoint can still connect while Cilium is not running`. It doesn't seem to be related to the new change.

**Comment by joestringer** [comment]
From what I can tell, we've had at least one failure of that particular test, and maybe five passes (GKE + net-next + 4.19 + 4.9 + test-focus) so it seems like even if it's an issue it's fairly rare. Looking at trends over the past two weeks, that test has mostly failed in GKE environments (most likely test code issue) + perhaps 6 failures across PRs. Unless we think that we've run that test and had it failed that many times on this PR, I'm guessing it's just an infrequent flake that's present already in the tree.

@anfernee in future, don't be afraid to file a new issue if it looks unrelated, and copy/paste the full output of the failure + attach the CI logs and link. You can find examples of this by [searching for the labels area/ci under issues](https://github.com/cilium/cilium/issues?q=is%3Aopen+is%3Aissue+label%3Aarea%2FCI). That way we can attempt to process the sysdump/artifacts and try to draw conclusions / assess common failures over time. With just a name of a test failure it's very difficult to make any kind of assessment or ensure that we fix any underlying issue (whether infra, testing or code) that causes it.

**Comment by joestringer** [comment]
Ample reviews, all green CI, I think this is good to merge.
